gk/fib.c

```diff
@@ -0,0 +1,1640 @@
+
+		ip_prefix_present = rte_lpm_is_rule_present(
+			ltbl->lpm, ntohl(ip_prefix_addr->ip.v4.s_addr),
+			ip_prefix_len, &fib_id);
+		if (!ip_prefix_present) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: delete an non-existent IP prefix (%s)\n",
+				ip_prefix);
+			return NULL;
+		}
+
+		ip_prefix_fib = &ltbl->fib_tbl[fib_id];
+	} else if (likely(ip_prefix_addr->proto == ETHER_TYPE_IPv6)) {
+		uint8_t fib_id;
+
+		ip_prefix_present = rte_lpm6_is_rule_present(
+			ltbl->lpm6, ip_prefix_addr->ip.v6.s6_addr,
+			ip_prefix_len, &fib_id);
+		if (!ip_prefix_present) {
+			RTE_LOG(WARNING, GATEKEEPER,
+				"gk: delete an non-existent IP prefix (%s)\n",
+				ip_prefix);
+			return NULL;
+		}
+
+		ip_prefix_fib = &ltbl->fib_tbl6[fib_id];
+	} else {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"gk: delete an IP prefix (%s) with unknown IP type %hu\n",
+			ip_prefix, ip_prefix_addr->proto);
+		return NULL;
+	}
+
+	if (ip_prefix_fib->ref_cnt != 1) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: delete a fib entry [ip prefix = %s] with ref_cnt = %d, and action = %d\n",
+			ip_prefix, ip_prefix_fib->ref_cnt,
+			ip_prefix_fib->action);
+		return NULL;
+	}
+
+	if (ip_prefix_addr->proto == ETHER_TYPE_IPv4) {
+		ret = rte_lpm_delete(ltbl->lpm,
+			ntohl(ip_prefix_addr->ip.v4.s_addr), ip_prefix_len);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: cannot remove the IP prefix %s from LPM table\n",
+				ip_prefix);
+			return NULL;
+		}
+	} else {
+		ret = rte_lpm6_delete(ltbl->lpm6,
+			ip_prefix_addr->ip.v6.s6_addr, ip_prefix_len);
+		if (ret < 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"gk: cannot remove the IP prefix %s from LPM table\n",
+				ip_prefix);
+			return NULL;
+		}
+	}
+
+	return ip_prefix_fib;
+}
+
+/*
+ * This function is called by del_fib_entry_locked().
+ * Notice that, it doesn't stand on its own, and it's only
+ * a construct to make del_fib_entry_locked() readable.
+ */
+static int
+del_grantor_fib_entry_locked(const char *ip_prefix,
+	struct ipaddr *ip_prefix_addr, struct gk_fib *ip_prefix_fib)
+{
+	int ret;
+	struct ip_flow *flow = &ip_prefix_fib->u.grantor.flow;
+	struct ether_cache *eth_cache =
+		ip_prefix_fib->u.grantor.eth_cache;
+
+	/*
+	 * The Grantor is a neighbor, and the prefix
+	 * is the Grantor. So, we need to release the @eth_cache
+	 * Ethernet header entry from the neighbor hash table.
+	 */
+	if (ip_prefix_fib->u.grantor.next_fib->action ==
+			GK_FWD_NEIGHBOR_BACK_NET) {
+		/* Find the FIB entry for the Grantor. */
+		struct neighbor_hash_table *neigh_ht;
+
+		if (flow->proto == ETHER_TYPE_IPv4 && flow->f.v4.dst ==
+				ip_prefix_addr->ip.v4.s_addr) {
+			/*
+			 * The Ethernet header cache entry is still
+			 * referenced by another prefix, so we cannot
+			 * release this FIB entry.
+			 */
+			if (eth_cache->ref_cnt != 1) {
+				RTE_LOG(ERR, GATEKEEPER,
+					"gk: delete a grantor neighbor fib entry [ip prefix = %s], however, its cached Ethernet header is referenced by other entries\n",
+					ip_prefix);
+				return -1;
```

Comment:
If `eth_cache->ref_cnt` is zero, there's a bug. If `eth_cache->ref_cnt` is greater than one, we just decrement it. If `eth_cache->ref_cnt` is one, we release it as well.

Given that other parts of the code have the same need to deal with references to `struct ether_cache`, this code should be encapsulated into a function. This new function can even deal with the IPv4 and IPv6 cases together. A place that also needs this code is `del_gateway_fib_entry_locked()`.